### DISCUSSION TITLE
Error with Flow Channel Scaling Factor

### DISCUSSION THREAD

**Discussion by ianmh02** [idaholab/moose#28430] [Q&A Modules: Thermal Hydraulics] [answered]
Hello, I am working on simulating a case where a pipe splits into 10 pipes, they go through a heated section, then recombine into 1 pipe. What I am trying to do I modify the scaling_factor_1phase for a flow channel 1 phase, however, whenever I define `scaling_factor_1phase`, I get a terminal error that says: "The variable 'rhoA' has already been added with a different scaling factor.", however, I do not have this defined elsewhere. Below is my current input file.

`# Define regen properties.
L_regen = 1
n_elems_regen = 10
D_regen = 0.005

[GlobalParams]
    initial_p = 1e6
    initial_T = 60
    initial_vel = 1.0
    initial_vel_x = 1
    initial_vel_y = 0
    initial_vel_z = 0

    fp = fp

    closures = simple_closures
    f = 0

    gravity_vector = '0 0 0'
[]

[FluidProperties]
    [fp]
      type = IdealGasFluidProperties
      gamma = 1.4
      molar_mass = 0.01
    []
[]

[Closures]
    [simple_closures]
        type = Closures1PhaseTHM
    []
[]

[Components]
    [inlet_1]
        type = InletMassFlowRateTemperature1Phase
        input = 'pipe_1:in'
        m_dot = 5
        T = 100
    []

    [pipe_1]
        type = FlowChannel1Phase
        position = '0 0 0'
        orientation = '0 1 0'
        length = 1
        n_elems = 10
        A = 1
        scaling_factor_1phase = '0.1 0.1 0.1'
    []

    [regen_hs]
        type = HeatStructureCylindrical
        position = '0 0 0'
        orientation = '0 1 0'
        length = ${L_regen}
        n_elems = ${n_elems_regen}
        names = 'block'
        widths = '${fparse D_regen / 2.}'
        solid_properties = 'steel'
        solid_properties_T_ref = '300'
        n_part_elems = 5
    []
    
    [total_pwr_regen]
        type = TotalPower
        power = 2000
    []
    
    [regen_heating]
        type = HeatSourceFromTotalPower
        hs = regen_hs
        regions = block
        power = total_pwr_regen
    []
    
    [regen_ht]
        type = HeatTransferFromHeatStructure1Phase
        flow_channel = pipe_1
        hs = regen_hs
        hs_side = outer
        P_hf = '${fparse pi * D_regen}'
    []

    [jct1]
        type = JunctionParallelChannels1Phase
        position = '0 1 0'
        connections = 'pipe_1:out pipe_2:in'
        volume = 1e-5
    []

    [pipe_2]
        type = FlowChannel1Phase
        position = '0 1 0'
        orientation = '0 1 0'
        length = 1
        n_elems = 10
        A = 1
        # scaling_factor_1phase = '0.1 0.1 0.1'
    []

    [outlet]
        type = Outlet1Phase
        input = 'pipe_2:out'
        p = 1e6
    []
[]

[Preconditioning]
    [pc]
        type = SMP
        full = true
    []
[]

[Executioner]
    type = Transient
    scheme = bdf2

    start_time = 0
    dt = 0.1
    num_steps = 5000

    abort_on_solve_fail = true

    solve_type = 'newton'
    line_search = 'basic'
    petsc_options_iname = '-pc_type'
    petsc_options_value = ' lu'

    nl_rel_tol = 1e-6
    nl_abs_tol = 1e-5

    nl_max_its = 5
    l_tol = 1e-4
[]

[Outputs]
    exodus = true

    [console]
        type = Console
        outlier_variable_norms = false
    []
    print_linear_residuals = false
[] `

Could this error be coming from the heat structure by any chance? When the heat structure is removed from the input file, I get no error, but when it is added, the error appears.

**Comment by joshuahansel** [comment, marked as answer]
The "different scaling factor" it refers to is the one used on your other channel (`pipe_2`), which is using a default value that is different from your 0.1 value in `pipe_1`, so you need to have these match.

**Reply by ianmh02** [reply]
If I had a system of pipes, would they all have to have the same scaling factor, or only the pipes on either side of the heated section?
I am curious because only the heated section splits into multiple pipes for better heat flow, but external to the heated sections, the piping is all in one pipe.

**Reply by joshuahansel** [reply]
Right now, all pipe components, heated or unheated, need to have the same scaling factor. This is because a given variable name like `rhoA` in the nonlinear system of equations may only have a single scaling factor. There is no ability right now to have that scaling factor differ by block/subdomain. For `CONSTANT MONOMIAL` variables like we use in THM's flow formulation, this might be possible, but for nodal variables, it wouldn't make since because nodes can be shared between subdomains.

What is your objective in modifying the scaling factor from the default?

**Reply by ianmh02** [reply]
Oh ok, I see what you mean. I am trying to modify the scaling factor specifically for heated sections because the total heat generated from the source needs to be distributed between a given number of pipes; without the scaling, the average surface temperature of the pipes do not make sense for the problem I am simulating. Is there another way this could potentially be done in MOOSE? And what would be considered as a `CONSTANT MONOMIAL` variable - would there be specific components used for that? Thank you for your help so far!

**Reply by joshuahansel** [reply]
Are you suggesting that the scaling factor is actually affecting the solution? To clarify, the scaling factor gets applied to the equations associated with that variable in the nonlinear system of equations, so in theory it should only affect

- conditioning of the linear system
- convergence criteria

So it should have numerical effects only, not effects to the governing equations.

If you are trying to change terms in the equations, then this is done elsewhere (let me know if this is the case). If you are not trying to modify terms in the equations and you are seeing solution differences with different scaling factors, it is likely because you are inadvertently changing convergence criteria.

So you have a 2D heat structure, and you want to split the heat between flow channels? What is the physical situation you are trying to model?

**Reply by ianmh02** [reply]
Yes, that is correct.; the model is currently set up for heat regen for a nuclear thermal engine. The pipes split off and wrap around the nozzle, gaining heat energy from conduction with the outer surface. We may try using a scaled up area approach for the time being to see how that works. Thank you for your help in clarifying!
